examples/nlp/language_modeling/megatron_gpt_distillation.py

```diff
@@ -0,0 +1,577 @@
+                            batch['loss_mask'], batch['num_valid_tokens_in_ub'], x
+                        )
+                    )
+                cp_size = parallel_state.get_context_parallel_world_size()
+                if validation_step and not self.validation_drop_last:
+                    num_valid_tokens_in_ub = batch['num_valid_tokens_in_ub']
+                    if loss_for_ub.isnan():
+                        assert batch['loss_mask'].count_nonzero() == 0, 'Got NaN loss with non-empty input'
+                        loss_sum_for_ub = torch.zeros_like(loss_for_ub)
+                        num_valid_tokens_in_ub = 0
+                    else:
+                        if self.sample_weight == 'constant':
+                            num_valid_tokens_in_ub = 1
+                        loss_sum_for_ub = num_valid_tokens_in_ub * loss_for_ub
+
+                    loss_sum_and_ub_size_all_gpu = torch.cat(
+                        [
+                            loss_sum_for_ub.clone().detach().view(1),
+                            torch.tensor([num_valid_tokens_in_ub]).cuda().clone().detach(),
+                        ]
+                    )
+                    # Could potentially reduce num_valid_samples_in_microbatch and use that to aggregate instead of len(self._validation_ds)
+                    torch.distributed.all_reduce(
+                        loss_sum_and_ub_size_all_gpu, group=parallel_state.get_data_parallel_group()
+                    )
+                    return loss_for_ub * cp_size, {'loss_sum_and_ub_size': loss_sum_and_ub_size_all_gpu}
+                else:
+                    reduced_loss = average_losses_across_data_parallel_group([loss_for_ub])
+                    return loss_for_ub * cp_size, {'avg': reduced_loss}
+
+            return output_tensor, loss_func
+
+        return fwd_output_and_loss_func
+
+    def loss_func(self, loss_mask, num_valid_tokens_in_ub, output_tensor, validation_step=False):
+        loss = super().loss_func(loss_mask, num_valid_tokens_in_ub, output_tensor)
+        if not validation_step and self.cfg.tensor_model_parallel_size > 1:
+            # [ModelOpt] KD loss requires extra all-reduce to ensure same values across MP-TP partitions.
+            loss = torch.sum(tensor_parallel.gather_from_tensor_model_parallel_region(loss.reshape(1)))
+        return loss
+
+    def configure_optimizers(self):
+        with self.model.hide_teacher_model():
+            return super().configure_optimizers()
+
+
+########################################################
+
+
+def load_distillation_config(student_cfg: TransformerConfig) -> Dict[str, Any]:
+    """Create a default distillation config for MCore GPT Models.
+
+    Args:
+        student_cfg: Model config for student model.
+    """
+    logit_pair = ("output_layer", "output_layer")  # logit module names for MCoreGPTModel
+    tp_enabled = student_cfg.tensor_model_parallel_size > 1
+
+    cfg = {
```

Comment:
Can these settings also moved to `config.yaml`?

Reply:
I would rather leave it here for now since the name of the logits layer is fixed for GPT models, and the TP thing is just a check.

Then user has fewer things to accidentally not touch.

Reply:
Discussed to move it to config once intermediate distillation support is added.